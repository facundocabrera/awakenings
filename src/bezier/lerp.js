/**
 * LERP
 *
 * @param {number} p0 real value
 * @param {number} p1 real value
 * @param {number} t  value between [0, 1]
 *
 * @returns Number
 */
const lerp = (p0, p1, t) => p0 * (1 - t) + p1 * t;

/**
 * LERP 2D
 *
 * @param {Array{2}} p0 [x1, y1]
 * @param {Array{2}} p1 [x2, y2]
 * @param {Number}   t  value between [0,1]
 *
 * @returns [x3, y3]
 */
const lerp2d = ([x1, y1], [x2, y2], t) => [
  /* x3 */ lerp(x1, x2, t),
  /* y3 */ lerp(y1, y2, t),
];

/**
 * Compute point p using bezier algorithm.
 *
 * @param {Array{n}} points points used as guide for the calculation
 * @param {Number}   t      value between [0,1]
 * @returns [x,y]
 */
const p = (points, t) => {
  if (points.length === 0) {
    throw Error("No points provided");
  }

  if (points.length === 1) {
    return points[0];
  } else {
    const ret = [];

    for (let i = 0; i < points.length - 1; i++) {
      ret.push(lerp2d(points[i], points[i + 1], t));
    }

    return p(ret, t);
  }
};

/**
 * Bezier 2D
 *
 * @param {Array<Array<Number,Number>>} points2d points used as guide for the calculation
 * @param {Number} time value between [0,1]
 * @returns [x,y]
 */
const bezier2d = (points2d, time) => points2d.reduce((prev, curr) => lerp2d(prev, curr, time));

export { lerp, lerp2d, p, bezier2d };
