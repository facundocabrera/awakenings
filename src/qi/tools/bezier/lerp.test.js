import { lerp, lerp2d, bezier2d } from "./lerp";

describe("LERP", () => {
  it("should do the math", () => {
    const p0 = 10;
    const p1 = 20;

    // t = 0 (min)
    expect(lerp(p0, p1, 0)).toBe(p0);
    // t = 1 (max)
    expect(lerp(p0, p1, 1)).toBe(p1);
  });

  it("calculate 2d lerp point", () => {
    const [x1, y1] = [-10, 10];
    const [x2, y2] = [10, 10];
    const t = 0.5;

    const x3 = lerp(x1, x2, t);
    const y3 = lerp(y1, y2, t);

    // es lo mismo que calcular el punto medio entre los 2 puntos iniciales
    expect([x3, y3]).toEqual(lerp2d([x1, y1], [x2, y2], t));
  });
});

describe("Bezier 2D", () => {
  const contract = (t, expected) => () => {
    const points = [
      [-10, 0],
      [0, 10],
      [10, 0],
    ];
    expect(bezier2d(points, t)).toEqual(expected);
  };

  it(
    "should build the bezier point of the curve at time t = 0",
    contract(0, [-10, 0])
  );
  it(
    "should build the bezier point of the curve at time t = 1",
    contract(1, [10, 0])
  );
  it(
    "should build the bezier point of the curve at time t = 0.5",
    contract(0.5, [0, 5])
  );
});
