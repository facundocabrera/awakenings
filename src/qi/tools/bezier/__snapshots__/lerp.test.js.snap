// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bezier2d sequencial reduce using lerp 2d 1`] = `
Array [
  25,
  125,
]
`;
